gzip.go

```diff
@@ -102,15 +166,17 @@ func (w *GzipResponseWriter) init() {
 	gzw := gzipWriterPools[w.index].Get().(*gzip.Writer)
 	gzw.Reset(w.ResponseWriter)
 	w.gw = gzw
-	w.ResponseWriter.Header().Set(contentEncoding, "gzip")
-	// if the Content-Length is already set, then calls to Write on gzip
-	// will fail to set the Content-Length header since its already set
-	// See: https://github.com/golang/go/issues/14975
-	w.ResponseWriter.Header().Del(contentLength)
 }
 
 // Close will close the gzip.Writer and will put it back in the gzipWriterPool.
 func (w *GzipResponseWriter) Close() error {
+	// Buffer not nil means the regular response it returned.
+	if w.buff != nil {
+		w.writeHeader()
+		w.ResponseWriter.Write(w.buff)
+		return nil
```

Comment:
Minor point: We don't need to return early here, because we'll do so immediately below, since w.gw is nil. I think it's better that way, since we don't have to rely on the (w.buff == nil) == (w.gw == nil) behavior never changing.